src/directories.cc

```diff
@@ -21,10 +23,87 @@ namespace sigc {
 #endif
 }
 
+bool Directories::TreeStore::row_drop_possible_vfunc(const Gtk::TreeModel::Path& path, const Gtk::SelectionData& selection_data) const {
+  return true;
+}
+
+bool Directories::TreeStore::drag_data_received_vfunc(const TreeModel::Path &path, const Gtk::SelectionData &selection_data) {
+  auto &directories=Directories::get();
+  
+  auto get_target_folder=[this, &directories](const TreeModel::Path &path) {
+    if(path.size()==1)
+      return directories.current_path;
+    else {
+      auto it=get_iter(path);
+      if(it) {
+        auto prev_path=path;
+        prev_path.up();
+        it=get_iter(prev_path);
+        if(it)
+          return it->get_value(directories.column_record.path);
+      }
+      else {
+        auto prev_path=path;
+        prev_path.up();
+        if(prev_path.size()==1)
+          return directories.current_path;
+        else {
+          prev_path.up();
+          it=get_iter(prev_path);
+          if(it)
+            return it->get_value(directories.column_record.path);
+        }
+      }
+    }
+    return boost::filesystem::path();
+  };
+  
+  auto it=directories.get_selection()->get_selected();
+  if(it) {
+    auto source_path=it->get_value(directories.column_record.path);
+    auto target_path=get_target_folder(path);
+    
+    target_path/=source_path.filename();
+    
+    if(source_path==target_path)
+      return false;
+    
+    if(boost::filesystem::exists(target_path)) {
+      Terminal::get().print("Error: Could not move file: "+target_path.string()+" already exists\n", true);
+      return false;
+    }
+    
+    boost::system::error_code ec;
+    boost::filesystem::rename(source_path, target_path, ec);
+    if(ec) {
+      Terminal::get().print("Error: Could not move file: "+ec.message()+'\n', true);
+      return false;
+    }
+    
+    for(int c=0;c<Notebook::get().size();c++) {
```

Comment:
Think it is best to do without. No need to support iterators when one never will use range algorithms (like std::equal) on the notebook tabs (compared to other tabs).